### DISCUSSION TITLE
Change PDF figure appearance in HTML format document

### DISCUSSION THREAD

**Discussion by nclJoshCowley** [quarto-dev/quarto-cli#3899] [Q&A] [answered]
I have a few HTML documents that use the default png figure format (images would be stored as `example_files/figure-html/*.png`)

I need to change this format to PDF when these images are to be used elsewhere (e.g. my Thesis)
For example,

````
---
title: "embed"
format: html
fig-format: pdf
---

```{r}
#| warning: false
ggplot2::qplot(1:10, (1:10)^2)
```
````

The output of this document leads to an embedded PDF in the HTML document.

That is, is has a scroll box instead of scaling to a readable size.

Is it possible to control how PDF images are displayed within a HTML document?

**Comment by mcanouil** [comment]
If you are using HTML, why do you want pdf figures?
I think svg would be better suited.

This being said your question is independent of Quarto CLI, you are more likely to ask/search using your favourite web search engine to find an answer.

**Comment by cscheid** [comment, marked as answer]
If you're producing a document with two separate formats, then it's probably a better idea to do something like

````
---
title: "embed"
format: 
  html: default
  pdf: default
---

```{r}
#| warning: false
ggplot2::qplot(1:10, (1:10)^2)
```
````

And now your HTML document will have bitmap (PNG) images, and your PDF document will have vector (PDF) images.

If you're not using quarto to produce your thesis, then you can do something like

````
```{r}
#| warning: false
fig <- ggplot2::qplot(1:10, (1:10)^2)
ggsave("out.pdf", fig, device="PDF")
fig
```
````

Which would produce "out.pdf" as a side effect of knitting your HTML document, while displaying the HTML document with a PNG image.

**Reply by mcanouil** [reply]
More backticks needed 😜

**Reply by baptiste** [reply]
Note that knits also lets you pass a vector of formats, as in

````
```{r, dev=c('png','pdf','svg')}
library(ggplot2)
ggplot() + ggtitle("this is a title")
```
````

so no need to worry about filename typos etc.

**Reply by cscheid** [reply]
@baptiste TIL! (I just verified this works with quarto as well.)

**Reply by mcanouil** [reply]
To avoid unnecessary code duplication, better to set this in the yaml header:

```yaml
knitr:
  opts_chunk:
    dev:
      - png
      - pdf
      - svg
```

---

Edit: I found an issue ^^

> Error in auto_exts[x] : invalid subscript type 'list'  
> Calls: .main ... process_group.block -> call_block -> block_exec -> eng_r -> dev2ext

The above does not work as it should, while below works fine:

````
```{r}
#| dev:
#|   - png
#|   - pdf
#|   - svg
library(ggplot2)
ggplot() + ggtitle("this is a title")
```
````

**Reply by cscheid** [reply]
> Error in auto_exts[x] : invalid subscript type 'list'
> Calls: .main ... process_group.block -> call_block -> block_exec -> eng_r -> dev2ext

@cderv Does the above look like a knitr bug to you? This is seemingly triggered by using a list of values in the `opts_chunk.dev` yaml key.

**Reply by nclJoshCowley** [reply]
I'm seeing this also, however, I just added the dev option to an already existing call to `opts_chunk$set`.

So using `knitr::opts_chunk$set(dev = c("png", "pdf", "svg"))` in a code chunk does work to reduce duplication.

**Reply by mcanouil** [reply]
@nclJoshCowley This is the knitr/rmarkdown syntax. The one not working is the Quarto/knitr syntax, which does not work when, in theory, it should.

**Reply by cderv** [reply]
@cscheid it seems setting the option globally works using `knitr::opts_chunk$set(dev = c("png", "pdf", "svg"))` in a chunk. 
The `opts_chunk` is a quarto way of doing that, so it seems there is something in quarto handling that could be missing - or some interaction in the order of options set maybe between knitr and quarto setting the option. 

I can have a closer look.

**Reply by baptiste** [reply]
By the way, there are a few related chunk options that follow the same logic (at least: ` dev, fig.ext, fig.width, fig.height, dpi, dev.args (list of lists)`)

**Reply by baptiste** [reply]
> @baptiste TIL! (I just verified this works with quarto as well.)

It's not widely known but a [decade-old in knitr](https://github.com/yihui/knitr/commit/cece3fb39b3f4295ad64401b163f5b8e206286ad); I guess my memory of it is helped by the fact that I appear to have been [pestering Yihui about it in the first place ;)](https://github.com/yihui/knitr/commit/20191a60356325e036760b1e99ac5f93b60a3b4a)

**Reply by mcanouil** [reply]
> By the way, there are a few related chunk options that follow the same logic (at least: ` dev, fig.ext, fig.width, fig.height, dpi, dev.args (list of lists)`)

But the others are working, I use `dev.args` all the time and no issue there, or at least I never found a case where it's not.

**Reply by baptiste** [reply]
I mean that they can all be vectorised (through knitr's syntax), as in 

````
```{r, dev=c('png','pdf'), fig.width=c(3,6),fig.height=c(3,3), dev.args=list(png=list(bg="red"), pdf=list(bg="green", family="Times"))}
plot(1,1)
title("this is the title")
```

````

(though I don't understand why the background is green and not red for the png, might be a bug/inconsistency with what's described at https://yihui.org/knitr/options/ )

**Reply by mcanouil** [reply]
The "green" might be an artefact from transparency.

**Reply by cderv** [reply]
> To avoid unnecessary code duplication, better to set this in the yaml header:
> 
> ```yaml
> knitr:
>   opts_chunk:
>     dev:
>       - png
>       - pdf
>       - svg
> ```
> 
> Edit: I found an issue ^^
> 
> > Error in auto_exts[x] : invalid subscript type 'list'
> > Calls: .main ... process_group.block -> call_block -> block_exec -> eng_r -> dev2ext
> 

So this was indeed an issue as the options are not process the same in-chunk and in YAML header. 
This should be now fixed in dev version of quarto (soon pre-release).

**Comment by nclJoshCowley** [comment]
Thanks both,

It is the case that I want a HTML version for viewing as there are a few tabsets to organise many plots.
The need for PDF is for vector graphics (SVG would also work) to be put into a LaTeX (non-quarto) document.

I think the multiple outputs in the YAML route would work best for me so marked @mcanouil as correct.
